dspace-api/src/main/resources/org/dspace/storage/rdbms/sqlmigration/postgres/V7.6_2023.09.18__workspaceitem_add_item_id_unique_constraint.sql

```diff
@@ -0,0 +1,9 @@
+--
+-- The contents of this file are subject to the license and copyright
+-- detailed in the LICENSE and NOTICE files at the root of the source
+-- tree and available online at
+--
+-- http://www.dspace.org/license/
+--
+
+ALTER TABLE workspaceitem ADD CONSTRAINT unique_item_id UNIQUE(item_id);
```

Comment:
Should we update this migration to ensure the data is currently correct?  In other words, if a site attempts to run this migration when they have an `item_id` listed multiple times in the `workspaceitem` table, then this migration will error out.

It may be better to see if there's a way to drop these duplicate entries **before** adding the constraint.  That will ensure this change doesn't make the upgrade more difficult.

Additionally, because this migration has been delayed, I think we need to change the name of these migration files to be a date in 2024.  Something like:
````
V7.6_2024.12.06__workspaceitem_add_item_id_unique_constraint.sql
````

Reply:
@tdonohue It indeed makes sense to me to try to repair the data. If we have multiple rows with conflicting values, I presume we want to try to replace them with a row that is most likely to be correct. I would like some feedback on which strategy would be best here.

## Disregard all but the first row
One possible strategy is to always pick the row with the smallest `workspace_item_id`, following the reasoning that any rows that were added later were added in error, and their data should be disregarded.
This leads to a fairly simple migration:

```SQL
with dedup as (
    select item_id, min(workspace_item_id) as workspace_item_id
    from workspaceitem
    group by item_id
    having count(workspace_item_id) > 1
)
delete from workspaceitem
using dedup
where workspaceitem.item_id = dedup.item_id and workspaceitem.workspace_item_id <> dedup.workspace_item_id;
```

## Disregard all but the last row
Same as above, but we pick the largest `workspace_item_id`, following the reasoning that the row that was added last is most likely to contain the most recent data.
Implementation is identical, except we use `max` instead of `min`.

## Per-column approach
We could also look at the semantics of each individual column, and try to decide how best to integrate the information we have from multiple rows.
For example, we could say that for the `page_reached` column, we should take the highest value, as that page must have been reached at some point (even if the rest of the data of that row is not accurate).

Here is a suggestion on how we could do that for each column:
- `workspace_item_id`: either assign a fresh ID, or just take the oldest ID (aggregate minimum).
- `multiple_titles`: aggregate with `bool_or`; alternatively, re-derive from metadata values (check if there are multiple `dc.title` values for the item).
- `multiple_files`: aggregate with `bool_or`; alternatively, re-derive from bitstreams (check if there are multiple bitstreams for the item).
- `published_before`: aggregate with `bool_or`.
- `stage_reached`: aggregate with `max`.
- `page_reached`: aggregate with `max`.
- `collection_id`: if these values disagree I am not sure we can do better than "arbitrarily pick one".

If we follow one interpretation of this scheme, we end up with the following query to produce the replacements of the duplicate values:

```SQL
select
  -- Pick the oldest one.
  min(workspace_item_id) as workspace_item_id,

  -- Straightforward aggregations.
  bool_or(multiple_titles) as multiple_titles,
  bool_or(published_before) as published_before,
  bool_or(multiple_files) as multiple_files,
  max(stage_reached) as stage_reached,
  max(page_reached) as page_reached,

  -- Column we're grouping by.
  item_id,

  -- Pick an arbitrary one, in this case always the first.
  -- Starting from Postgres 16, there's an `any_value` aggregation function, but for earlier versions we have to hack it with an array aggregate.
  -- Note that Postgres arrays are 1-indexed.
  ((array_agg(collection_id))[1]) as collection_id

from workspaceitem
group by item_id
having count(workspace_item_id) > 1
```

We would then migrate the table as follows:
```
-- Identify which rows have duplicates, and compute their replacements.
with dedup as (
    <previous query>
)

-- Delete ALL rows that have a duplicate (they will be replaced).
, del as (
    delete from workspaceitem
    using dedup
    where workspaceitem.item_id = dedup.item_id
)

-- Insert the replacements.
insert into workspaceitem
select * from dedup;
```

Reply:
I believe the current behavior is to throw an exception when multiple rows are returned (see [here](https://github.com/DSpace/DSpace/blob/dspace-8_x/dspace-api/src/main/java/org/dspace/content/dao/impl/WorkspaceItemDAOImpl.java#L81)).

So at the moment uniqueness is enforced for read operations but not for write operations.

This leaves us free to choose the strategy that semantically makes the most sense, as there is no existing DSpace behavior that we need to be compatible with.

Reply:
@KoenP : If DSpace is throwing an exception if multiple rows are returned, then that implies to me that no institutions running DSpace should have multiple rows (if they did, then an exception would be thrown and they'd want to fix it).

So, in that scenario, I think we should choose one of the easy options... either just disregard all but the first row or disregard all but the last row.  The behavior of these options should essentially be the same (as it will be extremely rare for a site to have multiple values), so I have no strong opinion which one you want to choose.

Reply:
@tdonohue Thanks for the feedback, I pushed the requested changes.